calyx-py/calyx/queue_call.py

```diff
@@ -96,35 +93,21 @@ def insert_main(prog, queue):
             [
                 read_cmd,  # Read `commands[i]`
                 write_cmd_to_reg,  # Write it to `cmd`
-                cb.par(  # Now, in parallel, act based on the value of `cmd`
-                    cb.if_(
-                        # Is this a pop?
-                        cmd_eq_0[0].out,
-                        cmd_eq_0[1],
-                        [  # A pop
-                            cb.invoke(  # First we call pop
-                                queue,
-                                in_cmd=cmd.out,
-                                ref_ans=ans,
-                                ref_err=err,
-                            ),
-                            # AM: my goal is that,
-                            # if err flag comes back raised,
-                            # we do not perform this write or this incr_j
-                            write_ans,
-                            incr_j,
-                        ],
-                    ),
-                    cb.if_(  # Is this a push?
-                        cmd_neq_0[0].out,
-                        cmd_neq_0[1],
-                        cb.invoke(  # A push
-                            queue,
-                            in_cmd=cmd.out,
-                            ref_ans=ans,
-                            ref_err=err,
-                        ),
-                    ),
+                cb.invoke(  # Call the queue with `cmd`
+                    queue,
+                    in_cmd=cmd.out,
+                    ref_ans=ans,
+                    ref_err=err,
+                ),
+                cb.if_(  # If it was a pop or a peek, write ans to the answer list
+                    cmd_le_1[0].out,
+                    cmd_le_1[1],
+                    [  # AM: I'd like to have an additional check hereL
+                        # if err flag comes back raised,
+                        # we do not perform this write_ans or this incr_j
+                        write_ans,
+                        incr_j,
+                    ],
```

Comment:
What's the intuition for why this cannot be in the `par` anymore?

Reply:
Ah, so the diff makes it looks like we went from `par (foo)` to `seq (foo)`, but actually we went from:
```
par (
  ( 
    if_case_1: action; additional_action_for_cases_1_and_2 
  ), 
  ( 
    if_case_2: action; additional_action_for_cases_1_and_2 
  ), 
  ( 
    if_case_3: action; 
  )
) # par ends 
```
to:
```
action;
if_case_1_or_2: additional_action_for_cases_1_and_2
```
And I think that's simpler.

Basically the `action` (in our specific case, in `invoke`) was the same in all the parallel branches, so I lifted it. Then the `additional_action_for_cases_1_and_2`, which in our specific case is `write_ans; incr_j`, is run after that. These cannot be run in parallel: the latter step depends on `ans`, which is passed by reference (and potentially modified) during the former step.

Reply:
Oh, gotcha! We should implement this kind of hoisting in a control simplification pass for Calyx. Maybe worth opening a short issue!